internal/controller/customreplicaset_controller.go

```diff
@@ -74,13 +80,130 @@ func (r *CustomReplicaSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
 	availableStdPods, availableUpgPods := countAvailablePods(childPods.Items)
 	totalAvailablePods := availableStdPods + availableUpgPods
 
+	// Load latest controller revision
+	var latestRevision *v1.ControllerRevision
+	if latestRevision, err = r.getLatestControllerRevision(ctx, &crs); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Apply any necessary updates to controller revision history
+	if err := r.manageControllerRevisionHistory(ctx, &crs, latestRevision); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+
 	if err := r.managePods(ctx, totalAvailablePods, availableStdPods, availableUpgPods, &crs, childPods, log); err != nil {
 		return ctrl.Result{Requeue: true}, err
 	}
 
 	return ctrl.Result{Requeue: false}, err
 }
 
+// Create a new ControllerRevision
+func (r *CustomReplicaSetReconciler) createControllerRevision(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet, latestRevision *v1.ControllerRevision, jsonSpec []byte) error {
+	var revisionNumber int64
+	if latestRevision == nil {
+		revisionNumber = 0
+	} else {
+		revisionNumber = latestRevision.Revision + 1
+	}
+
+	t := time.Now()
+	timestamp := fmt.Sprintf("%d%d%d%d", t.Hour(), t.Minute(), t.Second(), t.Nanosecond())
+
+	newRevision := v1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name + "-controller-revision-" + timestamp,
```

Comment:
You can slightly simplify the code here by doing:
{
    Name: fmt.Sprintf("..."),
}
If time struct has a built-in String() method, you can use it instead of writing your own in line 111.

Reply:
Awesome yep made this change to use simple `fmt.Sprintf("%s-controller-revision-%d", cr.Name, t.UnixNano())`